### PR TITLE
PY3: Fixed test_utils_python.py

### PR DESCRIPTION
**Pull request by klangner**

### CODE REVIEW

File: scrapy/utils/python.py [hunk 1]
```diff
@@ -58,9 +58,9 @@ def str_to_unicode(text, encoding=None, errors='strict'):
 
     if encoding is None:
         encoding = 'utf-8'
-    if isinstance(text, str):
+    if isinstance(text, six.binary_type):
```
Comment (kmike) [inline scrapy/utils/python.py]: six.binary_type is unnecessary since we don't support Python 2.6 - use `bytes`

**Comment by klangner** [comment]
ok. Done.

File: scrapy/utils/python.py [hunk 2]
```diff
@@ -122,14 +122,14 @@ def new_method(self, *args, **kwargs):
...
+_BINARYCHARS = set(map(six.int2byte, range(32))) - set([b"\0", b"\t", b"\n", b"\r"])
 
 def isbinarytext(text):
     """Return True if the given text is considered binary, or false
     otherwise, by looking for binary bytes at their chars
     """
-    assert isinstance(text, str), "text must be str, got '%s'" % type(text).__name__
-    return any(c in _BINARYCHARS for c in text)
+    assert isinstance(text, bytes), "text must be bytes, got '%s'" % type(text).__name__
+    return any(six.int2byte(c) in _BINARYCHARS for c in six.iterbytes(text))
```
Comment (kmike) [inline scrapy/utils/python.py]: Why isn't _BINARYCHARS a set of ints?

File: tests/test_utils_python.py [hunk 3]
```diff
@@ -1,6 +1,8 @@
 import functools
 import operator
 import unittest
+import six
```
Comment (kmike) [inline tests/test_utils_python.py]: A nitpick: I think `six` import should go after stdlib imports. pep8 quote:

```
 Imports should be grouped in the following order:

    standard library imports
    related third party imports
    local application/library specific imports

You should put a blank line between each group of imports. 
```

File: tests/test_utils_python.py [hunk 4]
```diff
@@ -24,23 +26,23 @@ def test_str_to_unicode(self):
...
         # converting a regular string to string should return the same object
-        self.assertEqual(unicode_to_str('lel\xf1e'), 'lel\xf1e')
+        self.assertEqual(unicode_to_str(b'lel\xf1e'), b'lel\xf1e')
 
         # converting a strange object should raise TypeError
         self.assertRaises(TypeError, unicode_to_str, unittest)
 
         # check errors argument works
-        assert '?' in unicode_to_str(u'a\ufffdb', 'latin-1', errors='replace')
+        self.assertEqual(unicode_to_str(u'a\ufffdb', 'latin-1', errors='replace'), b'a?b')
```
Comment (kmike) [inline tests/test_utils_python.py]: :+1:

File: scrapy/utils/python.py [hunk 5]
```diff
@@ -74,9 +74,9 @@ def unicode_to_str(text, encoding=None, errors='strict'):
...
     if encoding is None:
         encoding = 'utf-8'
-    if isinstance(text, unicode):
+    if isinstance(text, six.text_type):
         return text.encode(encoding, errors)
-    elif isinstance(text, str):
+    elif isinstance(text, bytes):
         return text
     else:
         raise TypeError('unicode_to_str must receive a unicode or str object, got %s' % type(text).__name__)
```
Comment (kmike) [inline scrapy/utils/python.py]: I think "unicode or str" should be changed to "unicode, str or bytes" or "unicode or bytes"

**Comment by klangner** [comment]
Ok. Fixed

**Comment by kmike** [comment]
@klangner sorry for being picky :) Could you please check the docstrings of the functions in this module? For example, `unicode_to_str` docstring says "Return the str representation" which is incorrect in Python 3. 

I think it is a good to review a module in more details while fixing the tests (and maybe add more tests) because we may think its porting is done when tests are passing while it is not.

**Comment by kmike** [comment]
re_rsearch from this module looks broken in Python 3.x, and as tests are passing it seems there is no test for it

**Comment by klangner** [comment]
There aren't any tests for re_search in python2. And it would be hard to add new tests while doing porting since I don't know what types this function expect. 
I would guess that sooner or later it will be necessary to fix this function because some other test will fail because of it. When I find this problem I will at least now what types are passed to this function (unicode or bytes).
And of course at the end when all tests will work it will be necessary to check if scrapy works with py3. If not it will be necessary to add new tests.

If I'm going to find all missing tests now than I'm afraid I want make any progress with porting.

**Comment by Curita** [comment]
Time-wise I think it's best to port full modules along with their tests, taking advantage that anyone reviewing tests would have some knowledge how their respective modules work. We'll have to do it anyway at some point, so doing so while reviewing that specific module's tests will alleviate the context-switch for fixing it.

It's true that we'll have to make full runs to catch some undiscovered errors, but we should try to fix some obvious ones beforehand.

Having said that, porting `re_rsearch` is not trivial since the `re` module has different behaviors across Python versions. You can't have different string types as *pattern* and *text* on Python 3 (but it's ok on Python 2) and compiling unicode regex expressions appends automatically the re.UNICODE flag on Python 3 (there aren't any additional flags on Python 2). 

Because of that, I'm ok with merging this PR as it is, but we should open a new issue to address it if this PR is not going to fix it.

**Comment by kmike** [comment]
Unfortunately we can't rely only on tests during the port. "Let's wait until it breaks and then fix it" doesn't sound good - it could break in user code, not in a Scrapy testing suite. We'll get a buggy Scrapy as a result, with issues hidden all over the codebase. Scrapy testing suite is not bad, but it doesn't cover all edge cases, and it doesn't cover all of Scrapy. 

Also, tests may become subtely incorrect in Python 3.x. For example, if we have a test which checks that `func() == "foo"` then in Python 3.x it means func() returns unicode, while in Python 2.x it means func() returns bytes or unicode (test will pass for both in 2.x). If func() should really return bytes, but we made it return unicode when porting, the tests will be still passing. This means we should think about each function; just making tests pass is not sufficient.

One extra problem is that tests can't catch errors in docs, comments and error messages. There is still a couple of such issues in this PR: 

* TypeError in `str_to_unicode` tells that an input can be either str or unicode, but in 3.x it can be either bytes or unicode;
* (not new) apparently, the docsting for `str_to_unicode` is incorrect - it should mention `.decode()` instead of `.encode()`.

**Comment by Curita** [comment]
I'm adding the easy tag so any GSoC students (specially those interested in the python 3 port) can follow up this work. Read full discussion first, and then create a new pull request based on this one. 

General idea is to add missing tests from `scrapy/utils/python.py` that will make evident that some functions break on python 3 (we know that at least `re_rseach` breaks under it), and then fix those functions so they work in both python versions. Search Scrapy's code to get a grip on how these functions work so you can make proper tests.

**Comment by kmike** [comment]
Thanks @klangner! Me, @eliasdorneles and @dangra went forward and merged https://github.com/scrapy/scrapy/pull/1379, copy-pasting some of your changes.
